Feature/ATG-106 Unit test

IraCherginyshka opened the pull request.

Commit:
unit test for Leader board
fix player index not in top

Commit:
Merge branches 'dev' and 'feature/ATG-106' of https://github.com/IraCherginyshka/AwesomeTetrisGame into feature/ATG-106

Commit:
pull dev

Commit:
mock all dependencies

Commit:
unit test: profile & profile-header component

Commit:
update yml file
add steps linter and test

Commit:
unit test for control-option component

Commit:
unit test for state-controls component

oleksandrstarov commented on .github/workflows/build-app.yml:
```diff
@@ -19,6 +19,12 @@ jobs:
         node-version: ${{ matrix.node-version }}
+    - name: linter
+      run: |
+        npm run eslint
```
the | symbol is in use for multiple commands and we have a single execution here, please, remove the | and write script at the same line (as it done on line 33 for example)

IraCherginyshka replied:
Deleted

oleksandrstarov commented on .github/workflows/build-app.yml:
```diff
@@ -19,6 +19,12 @@ jobs:
+        npm run eslint
+    - name: test
+      run: |
+        npm run test
```
same as above

IraCherginyshka replied:
Deleted

oleksandrstarov commented on .github/workflows/build-app.yml:
```diff
@@ -19,6 +19,12 @@ jobs:
       uses: actions/setup-node@v1
       with:
         node-version: ${{ matrix.node-version }}
+    - name: linter
```
Please note that on Git side we do not have any node_modules yet, the next steps will fail due to absence of linter and test packages, we have to first install npm packages and only after that we can run linter and tests

IraCherginyshka replied:
Fixed

oleksandrstarov commented on README.md:
```diff
@@ -43,11 +43,12 @@ $ npm run eslint
+`auto` steps:
+
+- checkout on `pr-pages` branch
+- run "ng build --prod",
```
Please update this file according to the changes in pipeline (lint and test steps)

IraCherginyshka replied:
Done

oleksandrstarov commented on README.md:
```diff
@@ -43,11 +43,12 @@ $ npm run eslint
+- checkout on `pr-pages` branch
+- run "ng build --prod",
+- copy `dist/` directory to the `server/` directory
+- delete all unnecessary files and directories
```
I would write delete all source files and folders that are not required for deployment

IraCherginyshka replied:
Changed

oleksandrstarov commented on package.json:
```diff
@@ -30,6 +31,7 @@
     "@angular/router": "~8.2.13",
     "eslint-config-prettier": "^6.5.0",
     "moment": "^2.24.0",
+    "ngx-mock-provider": "^1.0.6",
```
Please move this package to dev dependencies, as this is obviously not required for build, same for "eslint-config-prettier"

IraCherginyshka replied:
In the beginning, I added this package to dev dependencies and got the error
ESLint: 'ngx-mock-provider' should be listed in the project's dependencies, not devDependencies.(import/no-extraneous-dependencies)
What should I do?

"eslint-config-prettier" I moved.

IraCherginyshka replied:
I got the same error about ng2-mock-component: 
ESLint: 'ng2-mock-component' should be listed in the project's dependencies, not devDependencies.(import/no-extraneous-dependencies)

oleksandrstarov commented on src/app/app.component.spec.ts:
```diff
@@ -1,12 +1,30 @@
+        AppComponent,
+        GameHeaderComponent,
+        PlayerProfileHeaderComponent,
+        ToastContainerDirective,
```
I know this is done only to pass the testing build, but please rewrite this to use mocks instead of real components

IraCherginyshka replied:
Done

oleksandrstarov commented on src/app/components/game-control-option/game-control-option.component.spec.ts:
```diff
@@ -1,24 +1,105 @@
+    spyOn(component, 'saveControls').and.callThrough();
+
+    const button = fixture.debugElement.nativeElement.querySelector('#custom');
+    button.click();
```
This looks good, but it is enough to manually call component.saveControls() and observe the results

IraCherginyshka replied:
Done

oleksandrstarov commented on src/app/components/game-header/game-header.component.spec.ts:
```diff
@@ -1,20 +1,51 @@
     TestBed.configureTestingModule({
-      declarations: [GameHeaderComponent],
+      imports: [RouterTestingModule, BrowserAnimationsModule, ToastrModule.forRoot()],
+      declarations: [GameHeaderComponent, PlayerProfileHeaderComponent, ToastContainerDirective],
```
Same for component declarations - please use Mocks

IraCherginyshka replied:
Done

oleksandrstarov commented on src/app/components/game-leaderboard/game-leaderboard.component.html:
```diff
@@ -24,7 +24,7 @@
     <ul class="leaderboard">
       <li class="leaderboard__list">
-        <span class="leaderboard_index">{{ currentPlayerPlace }}</span>
+        <span class="leaderboard_index">{{ currentPlayerPlace + 1 }}</span>
```
I would rename variable to currentPlayerIndex to avoid confusion

IraCherginyshka replied:
Renamed

oleksandrstarov commented on src/app/components/game-leaderboard/game-leaderboard.component.ts:
```diff
@@ -24,20 +25,24 @@ export class GameLeaderboardComponent implements OnInit {
+
+      if (this.currentPlayerName) {
+        this.currentPlayerPlace = results.findIndex(
+          (player) => player.username === this.currentPlayerName,
```
can be shortened to ({ username }) => username === this.currentPlayerName,

IraCherginyshka replied:
Changed

oleksandrstarov commented on src/app/components/game-login-form/game-login-form.component.spec.ts:
```diff
@@ -1,24 +1,91 @@
+import { BrowserModule } from '@angular/platform-browser';
+import { FormsModule, ReactiveFormsModule } from '@angular/forms';
+import { RouterTestingModule } from '@angular/router/testing';
+import { RouterModule } from '@angular/router';
```
Not sure we need to have this module since we have RouterTestingModule

IraCherginyshka replied:
Deleted

oleksandrstarov commented on src/app/components/game-main/game-main.component.spec.ts:
```diff
@@ -1,20 +1,44 @@
+        GameBoardComponent,
+        GameNextFigureComponent,
+        GameStateControlsComponent,
+        GameControlsComponent,
```
Please mock these components

IraCherginyshka replied:
Done

Commit:
resolve conversation
fix steps in yml file // no | & npm ci before run linter
update readme file with new steps

Commit:
resolve conversation
delete test on click() //template
mock components
rename currentPlayerPlace => currentPlayerIndex
use destruction
delete RouterModule

IraCherginyshka committed:
Merge branch 'dev' into feature/ATG-106
